src/bin/stratisd.rs

```diff
@@ -365,30 +364,64 @@ fn run(matches: &ArgMatches, buff_log: &buff_log::Handle<env_logger::Logger>) ->
-                .iter()
-                .filter(|pfd| pfd.revents != 0)
-            {
-                for item in dbus_conn.watch_handle(pfd.fd, WatchEvent::from_revents(pfd.revents)) {
-                    if let Err(r) =
-                        libstratis::dbus_api::handle(&dbus_conn, &item, &mut tree, &dbus_context)
+            if let Some(ref mut handle) = dbus_handle {
+                for pfd in fds[dbus_client_index_start..]
+                    .iter()
+                    .filter(|pfd| pfd.revents != 0)
+                {
+                    for item in handle
+                        .connection
+                        .watch_handle(pfd.fd, WatchEvent::from_revents(pfd.revents))
                     {
-                        log_engine_state(&*engine.borrow());
-                        print_err(&From::from(r));
+                        if let Err(r) = libstratis::dbus_api::handle(
+                            &handle.connection,
+                            &item,
+                            &mut handle.tree,
+                            &handle.context,
+                        ) {
+                            log_engine_state(&*engine.borrow());
+                            print_err(&From::from(r));
+                        }
                     }
                 }
-            }
 
-            // Refresh list of dbus fds to poll for every time. This can change as
-            // D-Bus clients come and go.
-            fds.truncate(dbus_client_index_start);
+                // Refresh list of dbus fds to poll for. This can change as
+                // D-Bus clients come and go.
+                fds.truncate(dbus_client_index_start);
+                fds.extend(handle.connection.watch_fds().iter().map(|w| w.to_pollfd()));
+            } else {
+                // See if we can bring up dbus.
+                if let Ok(mut handle) = libstratis::dbus_api::connect(Rc::clone(&engine)) {
+                    info!("DBUS API is now available");
+
+                    // Register all the pools with dbus
+                    for (_, pool_uuid, pool) in engine.borrow().pools() {
+                        libstratis::dbus_api::register_pool(
+                            &handle.connection,
+                            &handle.context,
+                            &mut handle.tree,
+                            pool_uuid,
+                            pool,
+                            &handle.path,
+                        )?;
+                    }
 
-            fds.extend(dbus_conn.watch_fds().iter().map(|w| w.to_pollfd()));
+                    // Add dbus FD to fds as dbus is now available.
+                    fds.extend(handle.connection.watch_fds().iter().map(|w| w.to_pollfd()));
+                    dbus_handle = Some(handle);
+                }
+            }
         }
 
+        #[cfg(feature = "dbus_enabled")]
+        let poll_timeout = timeout(dbus_handle.is_none(), eventable.is_none());
```

Comment:
I'm guessing that all the arguments to timeout above and below have the wrong values, except for the literal ```false```. Assuming ```false``` is a correct argument, then it makes sense for the first argument to evaluate to ```false``` if D-Bus is _not_ enabled. So, in the call to timeout above, it makes more sense for the first argument to be ```dbus_handle.is_some()```, which will evaluate to true if stratisd has a D-Bus connection, but false otherwise. I think that a similar argument holds for the second argument.

Reply:
The arguments are correct and the function is doing the correct thing based on them.  However, I can see the confusion.  The `timeout` function simply selects the returned value in a prioritized order of shorter poll when dbus is compiled in but not available vs. an engine that isn't eventable.  When `dbus_handle.is_none()` is `true` dbus is not available, thus we want to the time out to be shorter, 1 second.  When `dbus_handle.is_none()` is true, then we evaluate if engine is eventable.  If `eventable.is_none()` is true then the engine isn't eventable and we want the timeout to be 10 seconds.  However, I need to ask @agrover why this was selected as the comments indicate that  when we aren't eventable that the call to the engine is a no-op.  Lastly when both dbus and eventable are both `false` we return -1 which is to wait forever on socket activity.

Obviously I need to re-work this to make it less confusing, so I'll see what I can do.  Passing in the reverse and negating the comparisons in `timeout` may be better.